### DISCUSSION TITLE
Retrieve previous route and/or route history in Next router

### DISCUSSION THREAD

**Discussion by michaeljblum** [vercel/next.js#36723] [Help] [answered]
I'm using the `router` you can retrieve from arguments in `_app`. 

In `router.route`, there is a state object identical with what is shown [here](https://nextjs.org/docs/api-reference/next/router) (docs).

Nowhere do I see any mention of past route retrieval. Is this not offered at all? Or is there still some Next way of doing this, other than explicit global state recording keeping or something?

**Comment by icyJoseph** [comment, marked as answer]
I think this is a tricky one... 

Generally, all routing solutions, use the history API, and that has its ups and downs. You can see in the browser yourself, `window.history` prints a few properties. 

Routing solutions take care of a bunch of things for you, but the idea of giving access to the history stack is terrifying, because, it is highly mutable (replace for example).

That being said, I can think of a couple of approaches:

```ts
// this would have to go in `_app.tsx` to be a true previous route tracker (across all pages)
const usePreviousRoute = () => {
  const { asPath } = useRouter();

  const ref = useRef<string | null>(null);

  useEffect(() => {
    ref.current = asPath;
  }, [asPath]);

  return ref.current;
};
```
> The above must be present on `_app.tsx` to track all pages.
> Remember that the initial state of previous route is null, or something else you decide.

You can also attach an event listener to `routeChangeComplete`, like `Router.events.on("routeChangeComplete", (url) => {})`, and keep track of it on a single element queue.

I'd also say that `asPath` is a bit better, because it includes the search (queries) of the URL.

At the top of my head, right now, there shouldn't be any issues with this, but perhaps you can find something? or confirm that it does the trick for you?

**Reply by michaeljblum** [reply]
Thanks! The asPath from useRouter() plus ref/useEffect approach indeed works for me. Having history or state like in reach router would be nice but this certainly works too!

**Reply by Starunski** [reply]
Thanks ! Also work for me

**Reply by wmonecke** [reply]
Thank you! In the end my hook looked something like this.

```javascript
import { useRouter } from 'next/router';
import { useRef } from 'react';

export const usePreviousRoute = () => {
  const router = useRouter();

  const ref = useRef<string | null>(null);

  router.events?.on('routeChangeStart', () => {
    ref.current = router.asPath;
  });

  return ref.current;
};
```

Then used the value via a context provider.

**Reply by cglacet** [reply]
With the app navigation, I tried this and it seems to work, I haven't tested it much yet :

```tsx
"use client";

import { usePathname } from "next/navigation";
import { PropsWithChildren, createContext, useEffect, useRef } from "react";

export const PreviousPathnameContext = createContext<string | undefined>(undefined);

export default function PreviousPathnameProvider({ children }: PropsWithChildren<{}>) {
    const pathname = usePathname();
    const ref = useRef<string>();

    useEffect(() => {
        ref.current = pathname;
    }, [pathname]);

    return (
        <PreviousPathnameContext.Provider value={ref.current}>
            {children}
        </PreviousPathnameContext.Provider>
    );
}
```

**Reply by kasir-barati** [reply]
IDK how it is working for you guys but for me it was just the current address 😢 If possible I would like to ask of you if you have any working example on GH so I can see how it is implemented @cglacet @wmonecke @icyJoseph

**Reply by cglacet** [reply]
@kasir-barati maybe because your page is reloading? How do you navigate between pages to test this?

**Reply by kasir-barati** [reply]
It is hard to answer that question but let say that user might use address bar to navigate, or click on a link, or we might redirect user with something like `router.back()`.

It is working right now though. lemme see if I can link a repo for it :)

**Reply by Stillonov** [reply]
Thanks. 
@cglacet Where do you use this solution? In layout?

**Reply by cglacet** [reply]
@Stillonov you have to put the provider at the root of your app (inside the render of your root layout), then you can use the context in any component.

**Reply by AhmedMuhammedElsaid** [reply]
this one fits my needs , ThanQ

**Comment by Standaa** [comment]
If you are looking to achieve the same thing because you want to redirect a user to the previously seen page after successfuul signin the above won't work as the path will be reset when navigating to your SignInController.
Here is an alternative below to tackle this scenario.
TLDR You can fetch the previously seen url from the headers' referrer (server side in `getServerSideProps`):

```
const SignInPage: TechTreePage<SignInPageProps> = ({ providers, csrfToken, previousRoute }) => (
  <SignInController providers={providers} csrfToken={csrfToken} previousRoute={previousRoute} />
);

export const getServerSideProps: GetServerSideProps = async context => {
  const providers = await getProviders();
  const csrfToken = await getCsrfToken(context);

  let previousRoute = context.req.headers?.referer || undefined;

  if (previousRoute === undefined || previousRoute === '/auth/signin') {
    previousRoute = '/MY_DEFAULT_ROUTE';
  }

  return {
    props: { providers, csrfToken, previousRoute }
  };
};
```

Then in the view:
```
onClick={() => signIn(provider.id, { callbackUrl: previousRoute })}
```

**Reply by david-vendel** [reply]
This is failing when I use browser back button, previousRoute contains current page value.
To fix this I have to save previousRoute to context when `previousRoute !== router.asPath` and then use this context value in my `Link`.

**Reply by david-vendel** [reply]
Also note that getServerSideProps [can't](https://nextjs.org/docs/messages/gssp-export) be used with static export.

**Comment by byteab** [comment]
### for anyone using Zustand
``` js
import { useRouter } from "next/router";
import { useEffect, useRef } from "react";
import { create } from "zustand";

interface AsPathStoreType {
  prevAsPath: string | undefined;
  currentAsPath: string | undefined;
}

const asPathStore = create<AsPathStoreType>((set) => ({
  prevAsPath: undefined,
  currentAsPath: undefined,
}));

/** use as a hook to get prevAsPath and currentAsPath*/
export const useAsPath = () => {
  return asPathStore((state) => state);
};

/** use everywhere you like */
export const getAsPath = () => {
  return asPathStore.getState();
};

/** Only use this in _app.tsx or root it's like a Provider */
export const useAsPathInitializer = () => {
  const { asPath } = useRouter();
  const { currentAsPath } = useAsPath();

  useEffect(() => {
    if (currentAsPath !== asPath) {
      asPathStore.setState((state) => ({
        ...state,
        currentAsPath: asPath,
        prevAsPath: currentAsPath,
      }));
    }
  }, [asPath, currentAsPath]);
};

```

**Comment by MigHerCas** [comment]
You can also use `router.back` method

Documentation: https://nextjs.org/docs/pages/api-reference/functions/use-router#routerback

**Reply by asouthern729** [reply]
True but I believe the entire app reloads.

**Reply by janvorwerk** [reply]
Nope, the app does not reload when calling `router.back()` : you can test it by adding a `data-iwashere` attribute to the body of your page in the developer tools and see that it remains there after navigation.

**Comment by sulvii** [comment]
How do you do it on the latest app router

**Comment by bimalstha23** [comment]
**If you're using Next.js with the new App Router**

```typescript
import { create } from 'zustand';
import { useEffect } from 'react';
import { usePathname } from 'next/navigation';

interface AsPathStoreType {
  prevAsPath: string | undefined;
  currentAsPath: string | undefined;
}

const asPathStore = create<AsPathStoreType>((set) => ({
  prevAsPath: undefined,
  currentAsPath: undefined,
}));

/** Hook to get prevAsPath and currentAsPath */
export const useAsPath = () => asPathStore((state) => state);

/** Utility function to get paths outside components */
export const getAsPath = () => asPathStore.getState();

/** Path initializer for App Router, should be used in root (e.g., layout.tsx or page.tsx) */
export const useAsPathInitializer = () => {
  const pathname = usePathname();
  const { currentAsPath } = useAsPath();

  useEffect(() => {
    if (currentAsPath !== pathname) {
      asPathStore.setState((state) => ({
        prevAsPath: state.currentAsPath,
        currentAsPath: pathname,
      }));
    }
  }, [pathname, currentAsPath]);
};

**Reply by Malin88** [reply]
U cannot use usePatchname() in layout.tsx component

**Reply by misaon** [reply]
@Malin88 https://github.com/shahradelahi/next-extra?tab=readme-ov-file#next-extrapathname

**Comment by RemyJouni** [comment]
The cleanest solution for me was using `sessionStorage` and `usePathname`







      
    const pathname = usePathname();

    //storing the pathnames when the value changes. 
    useEffect(() => {  
        storePathValues();
      }, [pathname]);

      function storePathValues() {
        const storage = globalThis?.sessionStorage;
        if (!storage) return;

        // Set the previous path as the value of the current path.
        const prevPath = storage.getItem("currentPath");
        

        storage.setItem("prevPath", prevPath ?? "/");
        // Set the current path value by looking at the browser's location object.
        storage.setItem("currentPath", globalThis.location.pathname);
      }



      // you can get the prevPath like this
     const prevPath = sessionStorage.getItem("prevPath");

**Comment by nabinkhair42** [comment]
## Server Component (Recommended for Redirects)

```tsx
import { redirect } from "next/navigation";
import { headers } from "next/headers";

export default async function AnalyticsPage() {
  const headersList = await headers();
  const referer = headersList.get("referer") || "";

  if (referer.includes("/analytics/pages")) {
    redirect("/analytics/pages");
  } else {
    redirect("/analytics/bots");
  }
}
```

---

## Client-Side Alternative (For State Persistence)

If you need reliable tracking across all navigation types, use `sessionStorage`:

```tsx
"use client";

import { usePathname } from "next/navigation";
import { useEffect } from "react";

export function useRouteHistory() {
  const pathname = usePathname();

  useEffect(() => {
    const current = sessionStorage.getItem("currentPath");
    if (current !== pathname) {
      sessionStorage.setItem("prevPath", current || "/");
      sessionStorage.setItem("currentPath", pathname);
    }
  }, [pathname]);

  return {
    current: pathname,
    previous:
      typeof window !== "undefined" ? sessionStorage.getItem("prevPath") : null,
  };
}
```

---

## TL;DR

| Use Case              | Solution                             |
| --------------------- | ------------------------------------ |
| Server-side redirects | `headers().get("referer")`           |
| Client-side tracking  | `usePathname()` + `sessionStorage`   |
| Need history stack    | Zustand/Context with `usePathname()` |
